src/provider/repository/mod.rs

```diff
@@ -0,0 +1,51 @@
+use std::convert::TryFrom;
+
+use kube::api::ListParams;
+use kube::{Api, Client};
+use log::{debug, trace};
+
+use crate::provider::error::StackableError;
+use crate::provider::repository::package::Package;
+use crate::provider::repository::repository::Repository;
+use crate::provider::repository::stackablerepository::StackableRepoProvider;
+
+pub mod package;
+pub mod repository;
+pub mod stackablerepository;
+
+pub async fn find_repository(
+    client: Client,
+    package: &Package,
+    repository_reference: Option<String>,
+) -> Result<Option<StackableRepoProvider>, StackableError> {
+    let repositories: Api<Repository> = Api::namespaced(client.clone(), "default");
+    if let Some(repository_name) = repository_reference {
+        // A repository name was provided, just check that exact repository for the package
+        let repo = repositories.get(&repository_name).await?;
+        let mut repo = StackableRepoProvider::try_from(&repo)?;
+        if repo.provides_package(package.clone()).await? {
+            return Ok(Some(repo));
+        }
+        return Ok(None);
+    } else {
+        // No repository name was provided, retrieve all repositories from the orchestrator/apiserver
+        // and check which one provides the package
+        let list_params = ListParams::default();
+        let repos = repositories.list(&list_params).await?;
+        for repository in repos.iter() {
+            let repo: &Repository = repository;
+            debug!("got repo definition: {:?}", repository);
+            // Convert repository to object implementing our trait
+            // TODO: add generic implementation here to support different types of repository
+            let mut repo = StackableRepoProvider::try_from(repository)?;
+            trace!("converted to stackable repo: {:?}", repository);
+            if repo.provides_package(package.clone()).await? {
```

Comment:
return can be lifted out

Reply:
Can it? We are only returning from on arm of the conditional, otherwise the code should continue with the next loop iteration..

Reply:
You're correct